packages/compiler/test/checker/resolve-type-reference-string.test.ts

```diff
@@ -0,0 +1,127 @@
+import { strictEqual } from "assert";
+import {
+  BasicTestRunner,
+  createTestHost,
+  createTestWrapper,
+  expectDiagnostics,
+} from "../../testing/index.js";
+
+describe("compiler: check: resolveTypeReferenceString", () => {
+  let runner: BasicTestRunner;
+  beforeEach(async () => {
+    runner = createTestWrapper(await createTestHost(), (x) => x);
+  });
+
+  async function expectResolve(reference: string, code: string) {
+    const { target } = await runner.compile(code);
+    if (target === undefined) {
+      throw new Error(`Must have @test("target") on type that should be referenced`);
+    }
+    const resolved = runner.program.checker.resolveTypeReferenceString(reference);
+    strictEqual(resolved, target);
+  }
+
+  async function diagnoseResolution(reference: string, code: string) {
+    await runner.compile(code);
+    const resolved = runner.program.checker.resolveTypeReferenceString(reference);
+    strictEqual(resolved, undefined);
+    return runner.program.diagnostics;
+  }
+
+  it("resolve simple namespace", async () => {
+    await expectResolve(
+      "MyService",
+      `
+      @test("target") namespace MyService {}
+    `
+    );
+  });
+
+  it("resolve nested namespace", async () => {
+    await expectResolve(
+      "MyOrg.MyService",
+      `
+      @test("target") namespace MyOrg.MyService {}
+    `
+    );
+  });
+
+  it("resolve model at root", async () => {
+    await expectResolve(
+      "Pet",
+      `
+      @test("target") model Pet {}
+    `
+    );
+  });
+
+  it("resolve model in namespace", async () => {
+    await expectResolve(
+      "MyOrg.MyService.Pet",
+      `
+      namespace MyOrg.MyService;
+      @test("target") model Pet {}
+    `
+    );
+  });
+
+  it("resolve model property", async () => {
+    await expectResolve(
+      "Pet.name",
+      `
+      model Pet { @test("target") name: string}
+    `
+    );
+  });
+
+  // Remove skip when this is added https://github.com/microsoft/cadl/issues/1092
+  it.skip("resolve metainfo", async () => {
+    await expectResolve(
+      "Pet.home::type.street",
+      `
+      model Pet { home: { @test("target") street: string}}
+    `
+    );
+  });
+
+  it("doesn't instantiate template", async () => {
```

Comment:
I'm confused by this test name

Reply:
check that if you are trying to reference a template instance it won't create a new type if it hasn't been instantiated yet.
(This function won't modify the type graph)

Reply:
Guess it doesn't check the type graph wasn't updated. Myabe I can check that but `diagnoseResolution` will check that it returned `undefined`. Could also maybe emit a dignostics in those cases